angr/state_plugins/fully_symbolic_memory.py

```diff
@@ -0,0 +1,972 @@
+
+        #op_start_time = time.time()
+        #print "Elapsed time: " + str(time.time() - op_start_time)
+
+        N = len(P)
+        merged_p = []
+        for i in range(N):
+
+            p = P[i]
+            v = p.obj
+
+            is_good_candidate = type(p.addr) == int and p.guard is None
+            mergeable = False
+
+            if len(merged_p) > 0 and is_good_candidate \
+                    and p.addr == merged_p[-1].addr + 1:
+
+                prev_v = merged_p[-1].obj
+                if v.op == 'BVV':
+
+                    # both constant and equal
+                    if prev_v.op == 'BVV' and v.args[0] == prev_v.args[0]:
+                        mergeable = True
+
+                # same symbolic object
+                elif v is prev_v:
+                    mergeable = True
+
+            if not mergeable:
+
+                if len(merged_p) > 0:
+                    obj = self.build_ite(addr, merged_p, merged_p[-1].obj, obj)
+                    merged_p = []
+
+                if is_good_candidate:
+                    merged_p.append(p)
+                else:
+                    obj = self.build_ite(addr, [p], v, obj)
+
+            else:
+                merged_p.append(p)
+
+        if len(merged_p) > 0:
+            obj = self.build_ite(addr, merged_p, merged_p[-1].obj, obj)
+
+        return obj
+
+    def _store_one_byte(self, obj, addr, k, min_addr, max_addr, condition):
+        if min_addr == max_addr:
+            assert addr == min_addr
+            P = self._concrete_memory[addr + k]
+            if P is None or condition is None:
+                self._concrete_memory[addr + k] = MemoryItem(addr + k, obj, self.timestamp,
+                                                             condition)
+            else:
+                item = MemoryItem(addr + k, obj, self.timestamp, condition)
+                if type(P) in (list,):
+                    P = [item] + P
+                else:
+                    P = [item, P]
+                self._concrete_memory[addr + k] = P
+        else:
+            # TODO: as in the paper, we should scan all memory objects that
+            # overlap with ours, and remove any which are equivalent
+            self._symbolic_memory.add(min_addr + k, max_addr + k + 1,
+                                      MemoryItem(addr + k, obj, self.timestamp, condition))
+
+    def _store_concrete_size(self, size, condition, data, addr, min_addr, max_addr):
+        # fastpath
+        # size is not symbolic so we don't have to chop data; MemoryItem lazily slices [data, k]
+        for k in range(size):
+            self._store_one_byte([data, k], addr, k, min_addr, max_addr, condition)
+        if condition is None:
+            return [ data ]
+        else:
+            return [ self.state.solver.If(condition, data, self.load(addr, size)) ]
+
+    def _store_symbolic_size(self, size, condition, data, addr, min_addr, max_addr, min_size, max_size):
```

Comment:
I split out the symbolic size and concrete size cases. Reason: req.stored_values wants us to give it sthg like

```
stored_values:
[0]: <BV8 if (size_21_32 > 0x0) then store_1 else orig_1>
[1]: <BV8 if (size_21_32 > 0x1) then store_2 else orig_2>
...
```

Chopping up these BV's is super slow for some reason. We only _have_ to do this for the symbolic size case (concrete size we can just return the data as-is). So, I split the two cases out.

Reply:
Do we actually _do_ anything with req.stored_values? SimMemory currently does this w/ the result:

```py
if request.completed and action is not None:
    action.actual_addrs = request.actual_addresses
    action.actual_value = action._make_object(request.stored_values[0]) # TODO
    ...
```

Reply:
So, stored_values are in fact used in order to construct simactions, as per the snippet, but that TODO, the problem is that there's not a very good way to actually recompose the multiple stored values into one action. Perhaps we should get an action per sub-store? From what I can tell the stores are split up for things like address concretization and also breaking the store across pages, maybe?

Reply:
Gotcha, I actually misunderstood this anyway. In the symbolic size case SimSymbolicMemory was concatenating all the individual byte stores _before_ returning (I didn't see that).

stored_values should still be fixed for when stores are concretized though.